AWS CNI chaining support

tgraf opened the pull request:
This PR introduces two new aspecs:

**Generic veth CNI plugin:**

This generic plugin identifies a veth peer in the container namespace and
extracts all relevant information from it, creates a Cilium endpoint and
instructs Cilium to install ingress and egress programs.

It can be used to build chaining with any veth-based CNI plugin.

**AWS CNI chaining mode:**

Building on top of the generic veth CNI plugin, an AWS CNI specific mode is introduced which written the appropriate CNI configuration file to chain aws-cni .+ portmap + Cilium.

<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/7862)
<!-- Reviewable:end -->

mumoshu commented on Documentation/gettingstarted/cni-chaining-aws-cni.rst:
```diff
@@ -0,0 +1,131 @@
+network devices as well as address allocation (IPAM) via ENI. After the initial
+networking is setup, the Cilium CNI plugin is called attach BPF programs to the
+network devices set up by aws-cni to enforce network policies, perform
+load-balancing, and encryption.
```
Does this integration allows us to apply CiliumNetworkPolicy containing labels?

My understanding was that this is basically [Native Routing](https://cilium.readthedocs.io/en/v1.5/intro/#simple-networking), so we are still limited to use security group IDs and pod/service CIDRs only for CiliumNetworkPolicy rule conditions. But just wanted to confirm.

tgraf replied:
@mumoshu This integration will allow you to use the full scope of CiliumNetworkPolicy including identity-based enforcement. This delegates the IPAM and network plumbing to aws-cni and uses Cilium to perform all policy and load-balancing aspects.

tgraf committed:
cni: Add ability for a chaining plugin to be called on delete

tgraf committed:
cni: Add support for AWS CNI chaining

Adds chaining plugin for AWS-CNI to CNI layer. The plugin is enabled via new
ConfigMap variable `cni-chaining-mode` which can be set to aws-cni.

Enabling chaining will automatically remove the configuration file
05-cilium.conf and replace it with 05-cilium.conflist and vice versa.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
cni: Add generic veth chaining plugin

This generic plugin identifies a veth peer in the container namespace and
extracts all relevant information from it, creates a Cilium endpoint and
instructs Cilium to install ingress and egress programs.

It can be used to build chaining with any veth-based CNI plugin.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
cni: Fix parsing of previous result

While testing with variour other plugins that provide different versions of the
CNI result, this allowed to work across various combinations.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
doc: Document aws-cni chaining mode

Signed-off-by: Thomas Graf <thomas@cilium.io>

aanm commented on plugins/cilium-cni/cilium-cni.go:
```diff
@@ -503,14 +503,45 @@ func cmdDel(args *skel.CmdArgs) error {
+			}
+		)
+
+		if chainAction.ImplementsDelete() {
```
What does it mean `ImplementsDelete`? How come a CNI plugin in chaining mode wouldn't implement delete?

tgraf replied:
There are plugins which do not have their own delete. Portmap is a good example.

aanm commented on plugins/cilium-cni/chaining/flannel/flannel.go:
```diff
@@ -141,6 +141,14 @@ func (f *flannelChainer) Add(ctx context.Context, pluginCtx chainingapi.PluginCo
 	return &cniTypesVer.Result{}, nil
 }
 
+func (f *flannelChainer) ImplementsDelete() bool {
```
Why doesn't flannel support delete?

tgraf replied:
Because there was no flannel specific delete code in the existing CNI plugin. This PR is not fixing up flannel. It is enabling chaining plugins to provide their own delete logic. We can convert flannel to be a fully native chaining plugin on top of this. I'm keeping the PR specific to enabling chaining for AWS-CNI.

aanm replied:
@tgraf maybe I'm missing something here. What's the purpose of this function? What do you mean by "enabling chaining for AWS-CNI"?

tgraf replied:
The function indicates whether a plugin has its own delete logic that should be called or whether the existing CNI DEL logic should be invoked. The new `aws-cni` chaining plugin has delete functionality, the existing `portmap` and `flannel` do not have own delete logic and continue to operate as before.

tgraf replied:
Discussed offline

aanm commented on plugins/cilium-cni/chaining/generic-veth/generic-veth.go:
```diff
@@ -0,0 +1,225 @@
+	return true
+}
+
+func (f *genericVethChainer) Add(ctx context.Context, pluginCtx chainingapi.PluginContext) (res *cniTypesVer.Result, err error) {
```
What's a "generic-veth"? Can't we have a name less "generic"?

tgraf replied:
It's a generic, hence the name :-), that supports any plugin that uses a veth pair. Let me know if you can think of a better name but it's literally a generic veth plugin.

tgraf replied:
Discussed offline

aanm commented on plugins/cilium-cni/chaining/generic-veth/generic-veth.go:
```diff
@@ -0,0 +1,225 @@
+	return true
+}
+
+func (f *genericVethChainer) Add(ctx context.Context, pluginCtx chainingapi.PluginContext) (res *cniTypesVer.Result, err error) {
```
What's a "generic-veth"? Can't we have a name less "generic"?

tgraf replied:
Discussed offline

aanm commented on plugins/cilium-cni/cni-install.sh:
```diff
@@ -108,6 +111,30 @@ EOF
+"aws-cni")
+	cat > ${CNI_CONF_NAME} <<EOF
+{
+  "name": "aws-cni",
```
I think we should start thinking on adding a `"version"` fields on these files.

tgraf replied:
Can you be more concrete? Specify the CNI version? That creates its own problems as the plugins we are chaining with may support different versions. The CNI code translates between versions so we can currently chain old and new versions.

aanm replied:
Not the CNI version but a version of that configuration file in order for us to understand. If we start adding more functionalities how will we know if future cilium-cni still understand's today's configuration or vice-versa.

tgraf replied:
We maintain 100% backward compatibility. What advantage does encoding a version have?

aanm commented on Documentation/gettingstarted/cni-chaining-aws-cni.rst:
```diff
@@ -0,0 +1,131 @@
+cluster or use any other method of your preference to set up a Kubernetes
+cluster.
+
+Ensure that the `aws-vpc-cni-k8s <https://github.com/aws/amazon-vpc-cni-k8s>`_
```
`aws-vpc-cni-k8s <https://github.com/aws/amazon-vpc-cni-k8s>`__

(Add 2 `_` at the end to make `aws-vpc-cni-k8s` a private reference otherwise we will have warnings if we create another href that uses the same name.)

aanm commented on Documentation/gettingstarted/cni-chaining-aws-cni.rst:
```diff
@@ -0,0 +1,131 @@
+
+      cni-chaining-mode: aws-cni
+      masquerade: "false"
+      tunnel: disabled
```
Hmm I didn't check the code, for flannel I enforce the tunnel to be disabled and print a warning in the logs in case the user forgot to do that.

tgraf replied:
What is your request? I do not want to make the agent aware of individual chaining modes. The agent is kept generic to support any plugin combination.

aanm replied:
Let me ask you differently, what are the consequences if the user does not set `disabled` in tunnel?

tgraf replied:
It will perform unnecessary tunneling. It will still work though.

aanm commented on Documentation/gettingstarted/cni-chaining-aws-cni.rst:
```diff
@@ -0,0 +1,131 @@
+
+Validate your AWS security groups rules and ensure that ENI IP addresses as
+allocated and used by the aws-cni plugin are allowed. See the documentation of
+the `aws-vpc-cni-k8s <https://github.com/aws/amazon-vpc-cni-k8s>`_ plugin for
```
same here for the `__`

aanm commented on Documentation/gettingstarted/cni-chaining-aws-cni.rst:
```diff
@@ -0,0 +1,131 @@
+As Cilium is deployed as a DaemonSet, it will write a new CNI configuration
+``05-cilium.conflist`` which will take precedence over the standard
+``10-aws.conflist``. Any new pod scheduled, will use the chaining configuration
+which will not also invoke Cilium.
```
maybe we should recommend deleting the `10-aws.conflist` to avoid the container runtime from using both?

tgraf replied:
That's likely easier said than done, it will get re-created on various occasions. I've added a note to the docs though.

aanm commented on Documentation/gettingstarted/cni-chaining-aws-cni.rst:
```diff
@@ -0,0 +1,131 @@
+load-balancing is not performed for traffic originating from existing pods.
+You must restart these pods in order to invoke the
+chaining configuration on them.
+
```
It is possible to know which pods are them, it's basically all pods that don't have a cilium endpoint in k8s and are not running in host network. We should make this as an opt-in in the operator for all pods and not only [kube|core]-dns.

tgraf replied:
I don't like the operator restarting application pods. I don't think users will want that. The majority of users will install Cilium upfront before running app pods so the notice is fine I think. I've added a note to the docs to help users identify pods that need to be restarted.

nebril commented on Documentation/gettingstarted/cni-chaining-aws-cni.rst:
```diff
@@ -0,0 +1,131 @@
+This guide explains how to set up Cilium in combination with aws-cni. In this
+hybrid mode, the aws-cni plugin is responsible for setting up the virtual
+network devices as well as address allocation (IPAM) via ENI. After the initial
+networking is setup, the Cilium CNI plugin is called attach BPF programs to the
```
s/is called attach/is called to attach
